### PR TITLE
major: Refactor God class to components 

### PR DESCRIPTION
**Pull request by astubbs**
Refactor the God class AbstractParallelEoSStreamProcessor into:

- function runner, worker pool, worker
- controller, rebalance handler, configuration validator, subscription interface
- state machine
- control loop

### Checklist

- [ ] Documentation (if applicable)
- [ ] Changelog
- [ ] Ordering of components in new classes

### CODE REVIEW

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ControlLoop.java [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+    private void clearCommitCommand() {
+        synchronized (commitCommand) {
+            if (commitCommand.get()) {
+                log.debug("Command to commit asap received, clearing");
+                this.commitCommand.set(false);
+            }
+        }
+    }
+
+    private <R> int retrieveAndDistributeNewWorkNew() {
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ControlLoop.java:324]: name

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/FunctionRunner.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+//        for (WorkContainer<K, V> workContainer : batch) {
+//            workContainer.setFuture(tuples);
+//        }
+    }
+
+    /**
+     * Run the supplied function.
+     */
+    // todo return value never used
+    private List<ParallelConsumer.Tuple<ConsumerRecord<K, V>, R>> runUserFunction(List<WorkContainer<K, V>> workContainerBatch) {
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/FunctionRunner.java:74]: break into steps?

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/PCWorker.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    SimpleMeterRegistry metricsRegistry = new SimpleMeterRegistry();
+
+    Timer userFunctionTimer = metricsRegistry.timer("user.function");
+
+    WorkQueue<K, V> workQueue = new WorkQueue<>();
+
+    PCWorkerPool<K, V, R> parentPool;
+
+    public void loop() {
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/PCWorker.java:28]: extract worker queues features to isolated branch

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/PCWorkerPool.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                .map(ignore -> new PCWorker<>(this))
+                .collect(Collectors.toList());
+    }
+
+    protected ThreadPoolExecutor createWorkerPool(int poolSize) {
+        ThreadFactory defaultFactory;
+        try {
+            defaultFactory = InitialContext.doLookup(options.getManagedThreadFactory());
+        } catch (NamingException e) {
+            log.debug("Using Java SE Thread", e);
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/PCWorkerPool.java:55]: don't log stack trace for normal operation

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/WorkMailbox.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static io.confluent.parallelconsumer.internal.FunctionRunner.MDC_WORK_CONTAINER_DESCRIPTOR;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static lombok.AccessLevel.PRIVATE;
+import static lombok.AccessLevel.PROTECTED;
+
+@Slf4j
+@RequiredArgsConstructor
+public class WorkMailbox<K, V> {
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/WorkMailbox.java:27]: rename Controller Mailbox?

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/WorkQueue.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package io.confluent.parallelconsumer.internal;
+
+import io.confluent.parallelconsumer.state.WorkContainer;
+
+import java.util.ArrayDeque;
+import java.util.List;
+import java.util.Queue;
+
+public class WorkQueue<K, V> {
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/WorkQueue.java:9]: not needed in this branch? (worker queues)

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/state/WorkContainer.java [hunk 7]
```diff
@@ -76,9 +77,9 @@
     @Getter
     private Optional<Boolean> maybeUserFunctionSucceeded = Optional.empty();
 
-    @Getter
-    @Setter(AccessLevel.PUBLIC)
-    private Future<List<?>> future;
+//    @Getter
+//    @Setter(PUBLIC)
+//    private Future<List<Object>> future;
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/state/WorkContainer.java:82]: delete

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java [hunk 8]
```diff
@@ -242,1064 +50,41 @@ protected AbstractParallelEoSStreamProcessor(ParallelConsumerOptions<K, V> newOp
...
-            return controlThreadFuture.get().isDone();
-        }
-    }
-
-    /**
-     * Optional ID of this instance. Useful for testing.
-     */
-    @Setter
-    @Getter
-    private Optional<String> myId = Optional.empty();
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/AbstractParallelEoSStreamProcessor.java:607]: go through deprecation?

**Review by astubbs — COMMENTED** [review]
draft

**Comment by eddyv** [comment]
Closing - Stale.

**Pull request closed without merging**